### PR TITLE
EAGLE2: general part [2]

### PR DESCRIPTION
**Pull request by yukavio**
This PR is a part of https://github.com/sgl-project/sglang/pull/1498. The original PR was split into smaller PRs to facilitate review. This PR should be merged after https://github.com/sgl-project/sglang/pull/2128.

### CODE REVIEW

File: examples/runtime/engine/EAGLE_offline_batch_inference.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import sglang as sgl
```
Comment (merrymercy) [inline examples/runtime/engine/EAGLE_offline_batch_inference.py:1]: 1. Move this to `sglang/test/srt/test_eagle.py`
2. Make it a unit test with `import unittest` and assert the output matches the one w/o speculative decoding.

File: examples/runtime/engine/EAGLE_offline_batch_inference.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    # Create an LLM.
+    llm = sgl.Engine(
+        model_path="meta-llama/Llama-2-7b-chat-hf",
+        draft_model_path="kavio/Sglang-EAGLE-llama2-chat-7B",
+        num_speculative_steps=3,
+        eagle_topk=4,
+        num_draft_tokens=16,
+        speculative_algorithm="EAGLE",
+        mem_fraction_static=0.70,
```
Comment (merrymercy) [inline examples/runtime/engine/EAGLE_offline_batch_inference.py:23]: Can we not change the `mem_fraction_static`?

File: examples/runtime/engine/EAGLE_offline_batch_inference.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    # Create a sampling params object.
+    sampling_params = {"temperature": 0, "max_new_tokens": 256}
+
+    # Create an LLM.
+    llm = sgl.Engine(
+        model_path="meta-llama/Llama-2-7b-chat-hf",
+        draft_model_path="kavio/Sglang-EAGLE-llama2-chat-7B",
+        num_speculative_steps=3,
+        eagle_topk=4,
+        num_draft_tokens=16,
```
Comment (merrymercy) [inline examples/runtime/engine/EAGLE_offline_batch_inference.py:21]: What are the corresponding names of these variables in the paper?

File: python/sglang/srt/layers/logits_processor.py [hunk 4]
```diff
@@ -33,6 +37,13 @@ class LogitsProcessorOutput:
     # The logprobs of the next tokens.     shape: [#seq, vocab_size]
     next_token_logprobs: torch.Tensor = None
 
+    # Used by speculative inference (eagle)
+    # The output of transformer layers
+    hidden_states: Optional[torch.Tensor] = None
+    # backup of next_token_logits when use cuda graph
+    # id(next_token_logits_bak) == id(next_token_logits)
+    next_token_logits_bak: Optional[torch.Tensor] = None
```
Comment (merrymercy) [inline python/sglang/srt/layers/logits_processor.py:45]: Why is this needed? Can you do a backup in the place that calls this function?

File: python/sglang/srt/layers/attention/__init__.py [hunk 5]
```diff
@@ -23,20 +26,26 @@ def init_cuda_graph_state(self, max_bs: int):
...
         self,
         bs: int,
+        num_token: int,
         req_pool_indices: torch.Tensor,
         seq_lens: torch.Tensor,
-        encoder_lens: Optional[torch.Tensor] = None,
+        encoder_lens: torch.Tensor = None,
+        spec_info: "SpecInput" = None,
+        is_draft_runner: bool = False,
+        forward_batch: ForwardBatch = None,
```
Comment (merrymercy) [inline python/sglang/srt/layers/attention/__init__.py:35]: We do not want to pass in big objects such as `spec_info` and `forward_batch`. They contain too many fields, making it hard to reason which tensors are needed for cuda graph.
cuda graph is error-prone so we want to make the dependency very clear.

File: python/sglang/srt/layers/attention/flashinfer_backend.py [hunk 6]
```diff
@@ -124,19 +129,46 @@ def __init__(self, model_runner: ModelRunner):
...
+            self.forward_metadata = (use_ragged, extend_no_prefix, None)
         else:
             prefix_lens = forward_batch.extend_prefix_lens
 
             # Some heuristics to check whether to use ragged forward
-            if forward_batch.extend_num_tokens >= 4096 and self.num_wrappers == 1:
+            if (
+                forward_batch.extend_num_tokens >= 4096
+                and self.num_wrappers == 1
+                and not forward_batch.forward_mode.is_spec_verify()
```
Comment (merrymercy) [inline python/sglang/srt/layers/attention/flashinfer_backend.py:170]: this line `not forward_batch.forward_mode.is_spec_verify()` is always true.

File: python/sglang/srt/layers/attention/flashinfer_backend.py [hunk 7]
```diff
@@ -164,26 +197,58 @@ def init_cuda_graph_state(self, max_bs: int):
             cuda_graph_kv_indices.clone() for _ in range(self.num_wrappers - 1)
         ]
 
+        self.cuda_graph_custom_mask = torch.zeros(
```
Comment (merrymercy) [inline python/sglang/srt/layers/attention/flashinfer_backend.py:200]:
```suggestion
        # For verifying draft tokens
        self.cuda_graph_custom_mask = torch.zeros(
```

File: python/sglang/srt/layers/attention/flashinfer_backend.py [hunk 8]
```diff
@@ -218,11 +286,16 @@ def get_cuda_graph_seq_len_fill_value(self):
     def forward_extend(
         self, q, k, v, layer: RadixAttention, forward_batch: ForwardBatch
     ):
-        prefill_wrapper_paged = self.prefill_wrappers_paged[
-            self._get_wrapper_idx(layer)
-        ]
+        if forward_batch.forward_mode.is_spec_verify():
```
Comment (merrymercy) [inline python/sglang/srt/layers/attention/flashinfer_backend.py:289]: ForwardMode.SPEC_VERIFY.is_decode() is true, but why does it go to `forward_extend`?
We need a better classification of these modes. If you cannot easily distinguish them, please do not use combinations like
```
    def is_decode(self):
        return self in (ForwardMode.DECODE, ForwardMode.SPEC_VERIFY)
```

File: python/sglang/srt/layers/attention/flashinfer_backend.py [hunk 9]
```diff
@@ -164,26 +197,58 @@ def init_cuda_graph_state(self, max_bs: int):
...
+        self.cuda_graph_qo_indptr = [x.clone() for x in self.kv_indptr]
+
     def init_forward_metadata_capture_cuda_graph(
         self,
         bs: int,
+        num_token: int,
         req_pool_indices: torch.Tensor,
         seq_lens: torch.Tensor,
         encoder_lens: torch.Tensor = None,
+        spec_info: SpecInput = None,
```
Comment (merrymercy) [inline python/sglang/srt/layers/attention/flashinfer_backend.py:215]: spec_info is almost not used. You only check if it is a None. We can simplify this.

File: python/sglang/srt/layers/attention/flashinfer_backend.py [hunk 10]
```diff
@@ -436,36 +526,63 @@ def call_begin_forward(
...
-            kv_indices,
-            self.kv_last_page_len[:bs],
-            self.num_qo_heads,
-            self.num_kv_heads,
-            self.head_dim,
-            1,
-            data_type=self.data_type,
-            q_data_type=self.q_data_type,
-        )
+        if forward_batch is not None and forward_batch.forward_mode.is_spec_verify():
```
Comment (merrymercy) [inline python/sglang/srt/layers/attention/flashinfer_backend.py:558]: These if/else are very confusing. SPEC_VERIFY uses the FlashInferIndicesUpdater**Decode** but it uses Batch**Prefill**WithPagedKVCacheWrapper. It can be called with `forward_extend` and  `SPEC_VERIFY.is_decode()` is true.

Can you make it clear with a doc string, or can you create more classes to make it clean?

File: python/sglang/srt/managers/data_parallel_controller.py [hunk 11]
```diff
@@ -96,7 +96,7 @@ def __init__(self, server_args, port_args) -> None:
             else:
                 # This port is checked free in PortArgs.init_new.
                 # We hold it first so that the next dp worker gets a different port
-                sockets.append(bind_port(tmp_port_args.nccl_port))
+                sockets.append(bind_port(tmp_port_args.nccl_port[0]))
```
Comment (merrymercy) [inline python/sglang/srt/managers/data_parallel_controller.py:99]: Let us add a new filed `nccl_port_drat_model` instead of making it a list.

File: python/sglang/srt/managers/schedule_batch.py [hunk 12]
```diff
@@ -483,6 +491,10 @@ class ScheduleBatch:
     # device
     device: str = "cuda"
 
+    # speculative decoding
+    spec_info: SpecInput = None
```
Comment (merrymercy) [inline python/sglang/srt/managers/schedule_batch.py:495]: style

```suggestion
    spec_info: Optional[SpecInput] = None
```

File: python/sglang/srt/managers/schedule_batch.py [hunk 13]
```diff
@@ -483,6 +491,10 @@ class ScheduleBatch:
...
+    spec_algorithm: SpeculativeAlgorithm = SpeculativeAlgorithm.NONE
```
Comment (merrymercy) [inline python/sglang/srt/managers/schedule_batch.py:496]: we can use `Optional[SpeculativeAlgorithm] = None` here

File: python/sglang/srt/model_executor/forward_batch_info.py [hunk 14]
```diff
@@ -67,21 +93,49 @@ def is_prefill(self):
         return self == ForwardMode.PREFILL
 
     def is_extend(self):
-        return self == ForwardMode.EXTEND or self == ForwardMode.MIXED
+        return self in (
+            ForwardMode.EXTEND,
+            ForwardMode.MIXED,
+            ForwardMode.SPEC_EXTEND,
```
Comment (merrymercy) [inline python/sglang/srt/model_executor/forward_batch_info.py:99]: This seems confusing. We can discuss how to classify them or create separate if/else conditions for them.

File: python/sglang/srt/model_executor/forward_batch_info.py [hunk 15]
```diff
@@ -138,6 +192,10 @@ class ForwardBatch:
     token_to_kv_pool: BaseTokenToKVPool = None
     attn_backend: AttentionBackend = None
 
+    # Speculative decoding
+    spec_info: SpecInput = None
+    spec_algorithm: SpeculativeAlgorithm = SpeculativeAlgorithm.NONE
```
Comment (merrymercy) [inline python/sglang/srt/model_executor/forward_batch_info.py:197]: style
```suggestion
    spec_info: Optional[SpecInput] = None
    spec_algorithm: Optional[SpeculativeAlgorithm] = None
```

File: python/sglang/srt/server_args.py [hunk 16]
```diff
@@ -135,6 +137,18 @@ class ServerArgs:
...
+    draft_model_path: Optional[str] = None
+    speculative_algorithm: SpeculativeAlgorithm = None
+    num_speculative_steps: int = None
+    # should been set as 2^n
+    num_draft_tokens: int = None
+    # should been set as [1, 2, 4, 8]
+    eagle_topk: int = None
+    # should not been set by cli, it is only a placeholder
+    # which would be set and used in model_runner
+    draft_runner_cache_size: int = None
```
Comment (merrymercy) [inline python/sglang/srt/server_args.py:150]: do not add it here if it will be set later. Use `global_server_args_dict` instead

File: python/sglang/srt/managers/scheduler.py [hunk 17]
```diff
@@ -555,15 +571,6 @@ def handle_generate_request(
                 req.origin_input_ids_unpadded, req.image_inputs
             )
 
-            if len(req.origin_input_ids) > self.max_req_input_len:
```
Comment (merrymercy) [inline python/sglang/srt/managers/scheduler.py:558]: do not delete this.

File: python/sglang/srt/managers/scheduler.py [hunk 18]
```diff
@@ -186,6 +187,21 @@ def __init__(
...
+        if self.server_args.speculative_algorithm.is_not_none():
+            self.draft_worker = spec_worker_factory.get(
+                self.server_args.speculative_algorithm
+            )(
+                gpu_id=gpu_id,
+                tp_rank=tp_rank,
+                server_args=server_args,
+                nccl_port=port_args.nccl_port,
+                target_worker=self.tp_worker,
+                dp_rank=dp_rank,
```
Comment (merrymercy) [inline python/sglang/srt/managers/scheduler.py:200]: you can pass `draft_runner_cache_size` here instead of using `server_args`

File: python/sglang/srt/managers/scheduler.py [hunk 19]
```diff
@@ -909,13 +928,17 @@ def update_running_batch(self):
...
-
         if self.is_generation:
-            model_worker_batch = batch.get_model_worker_batch()
             if batch.forward_mode.is_decode() or batch.extend_num_tokens != 0:
-                logits_output, next_token_ids = self.tp_worker.forward_batch_generation(
-                    model_worker_batch
-                )
+                if self.server_args.speculative_algorithm.is_not_none():
+                    logits_output, next_token_ids, model_worker_batch = (
+                        self.draft_worker.forward_batch_speculative_generate(batch)
```
Comment (merrymercy) [inline python/sglang/srt/managers/scheduler.py:935]: Can you let `forward_batch_speculative_generate` take `model_worker_batch` as inputs?
For anything you need, please copy it to `model_worker_batch`. We need this style for the overlap scheduler.

File: python/sglang/srt/managers/tp_worker.py [hunk 20]
```diff
@@ -141,12 +151,17 @@ def forward_batch_generation(
...
         forward_batch = ForwardBatch.init_new(model_worker_batch, self.model_runner)
         logits_output = self.model_runner.forward(forward_batch)
         if launch_done:
             launch_done.set()
-        next_token_ids = self.model_runner.sample(logits_output, model_worker_batch)
+        if skip_sample:
+            next_token_ids = None
+        else:
+            next_token_ids = self.model_runner.sample(logits_output, model_worker_batch)
+        model_worker_batch.spec_info = forward_batch.spec_info
```
Comment (merrymercy) [inline python/sglang/srt/managers/tp_worker.py:164]: This assignment should not be there.
The information flow should be `ScheduleBatch -> ModelWorkerBatch -> ForwardBatch`.
We will need this style to make all things work in overlap mode.

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: python/sglang/srt/model_executor/forward_batch_info.py [hunk 21]
```diff
@@ -56,6 +78,10 @@ class ForwardMode(IntEnum):
     DECODE = auto()
     # Contains both EXTEND and DECODE when doing chunked prefill.
     MIXED = auto()
+    # Speculative Verify stage
+    SPEC_VERIFY = auto()
+    # Speculative draft Extend stage which after verify stage
+    SPEC_EXTEND = auto()
```
Comment (merrymercy) [inline python/sglang/srt/model_executor/forward_batch_info.py:84]: To make it clean whether it is draft or target model

```
SPEC_VERIFY -> TARGET_VERIFY
SPEC_EXTEND -> DRAFT_EXTEND
```
